### ISSUE TITLE
DatabaseCorruptError when tagging many mails as spam.

### ISSUE THREAD

**Issue by moddx** [astroidmail/astroid#275]
I  am consistently experiencing a hard crash, when marking quite a few mails (like 24) and then toggling their spam status (to active). This is what astroid writes to console.

I am using version 0.7 from the Arch User Repository.

```
[22:18:10.709364] [0x00007fe5fbd7d340] [debug] ql: loaded 2700 threads.
[22:18:10.710046] [0x00007fe5d37fe700] [info] db: closing db.
[22:18:10.710095] [0x00007fe5d37fe700] [debug] db: ro: waiting for lock to close..
[22:18:10.710112] [0x00007fe5d37fe700] [debug] db: ro: closing..
[22:18:31.975759] [0x00007fe5fbd7d340] [info] mw: starting multi key.
[22:18:33.867668] [0x00007fe5fbd7d340] [debug] tl: m k h
[22:18:33.874248] [0x00007fe5e197f700] [info] db: open db read-write.
[22:18:33.874328] [0x00007fe5e197f700] [debug] db: rw-s: waiting for rw lock.. (r-o open: 0)
[22:18:33.874345] [0x00007fe5e197f700] [debug] db: rw-s lock acquired.
[22:18:33.874777] [0x00007fe5e197f700] [debug] db: open time: 2.344 ms.
[22:18:33.874810] [0x00007fe5e197f700] [debug] toggle_action: threadid:0000000000000198
[22:18:33.874840] [0x00007fe5e197f700] [debug] nm (0000000000000198): add tag: spam
[22:18:33.876924] [0x00007fe5e197f700] [debug] toggle_action: threadid:0000000000000197
[22:18:33.876971] [0x00007fe5e197f700] [debug] nm (0000000000000197): add tag: spam
[22:18:33.880412] [0x00007fe5e197f700] [debug] toggle_action: threadid:0000000000000196
[22:18:33.880458] [0x00007fe5e197f700] [debug] nm (0000000000000196): add tag: spam
[22:18:33.884633] [0x00007fe5e197f700] [debug] toggle_action: threadid:0000000000000192
[22:18:33.884669] [0x00007fe5e197f700] [debug] nm (0000000000000192): add tag: spam
[22:18:33.886374] [0x00007fe5e197f700] [debug] toggle_action: threadid:0000000000000193
[22:18:33.886408] [0x00007fe5e197f700] [debug] nm (0000000000000193): add tag: spam
[22:18:33.890055] [0x00007fe5e197f700] [debug] toggle_action: threadid:0000000000000191
[22:18:33.890089] [0x00007fe5e197f700] [debug] nm (0000000000000191): add tag: spam
[22:18:33.894263] [0x00007fe5e197f700] [debug] toggle_action: threadid:0000000000000195
[22:18:33.896332] [0x00007fe5e197f700] [debug] nm (0000000000000195): add tag: spam
[22:18:33.898082] [0x00007fe5e197f700] [debug] toggle_action: threadid:0000000000000190
[22:18:33.898140] [0x00007fe5e197f700] [debug] nm (0000000000000190): add tag: spam
[22:18:33.902262] [0x00007fe5e197f700] [debug] toggle_action: threadid:0000000000000194
[22:18:33.902316] [0x00007fe5e197f700] [debug] nm (0000000000000194): add tag: spam
[22:18:33.906168] [0x00007fe5e197f700] [debug] toggle_action: threadid:000000000000018f
[22:18:33.906205] [0x00007fe5e197f700] [debug] nm (000000000000018f): add tag: spam
[22:18:33.907765] [0x00007fe5e197f700] [debug] toggle_action: threadid:000000000000019e
[22:18:33.907800] [0x00007fe5e197f700] [debug] nm (000000000000019e): add tag: spam
[22:18:33.909304] [0x00007fe5e197f700] [debug] toggle_action: threadid:000000000000019d
[22:18:33.909338] [0x00007fe5e197f700] [debug] nm (000000000000019d): add tag: spam
[22:18:33.910940] [0x00007fe5e197f700] [debug] toggle_action: threadid:000000000000019c
[22:18:33.910990] [0x00007fe5e197f700] [debug] nm (000000000000019c): add tag: spam
[22:18:33.912750] [0x00007fe5e197f700] [debug] toggle_action: threadid:000000000000018c
[22:18:33.912823] [0x00007fe5e197f700] [debug] nm (000000000000018c): add tag: spam
[22:18:33.920142] [0x00007fe5e197f700] [debug] toggle_action: threadid:000000000000018d
[22:18:33.920192] [0x00007fe5e197f700] [debug] nm (000000000000018d): add tag: spam
[22:18:33.922028] [0x00007fe5e197f700] [debug] toggle_action: threadid:000000000000019b
[22:18:33.922063] [0x00007fe5e197f700] [debug] nm (000000000000019b): add tag: spam
[22:18:33.923665] [0x00007fe5e197f700] [debug] toggle_action: threadid:000000000000018b
[22:18:33.923700] [0x00007fe5e197f700] [debug] nm (000000000000018b): add tag: spam
[22:18:33.924938] [0x00007fe5e197f700] [debug] toggle_action: threadid:000000000000019a
[22:18:33.924972] [0x00007fe5e197f700] [debug] nm (000000000000019a): add tag: spam
[22:18:33.926225] [0x00007fe5e197f700] [debug] toggle_action: threadid:000000000000018a
[22:18:33.926258] [0x00007fe5e197f700] [debug] nm (000000000000018a): add tag: spam
[22:18:33.927805] [0x00007fe5e197f700] [debug] toggle_action: threadid:0000000000000189
[22:18:33.927840] [0x00007fe5e197f700] [debug] nm (0000000000000189): add tag: spam
[22:18:33.929249] [0x00007fe5e197f700] [debug] toggle_action: threadid:0000000000000188
[22:18:33.929293] [0x00007fe5e197f700] [debug] nm (0000000000000188): add tag: spam
[22:18:33.930579] [0x00007fe5e197f700] [debug] toggle_action: threadid:0000000000000199
[22:18:33.932902] [0x00007fe5e197f700] [debug] nm (0000000000000199): add tag: spam
[22:18:33.934540] [0x00007fe5e197f700] [debug] toggle_action: threadid:0000000000000187
[22:18:33.934609] [0x00007fe5e197f700] [debug] nm (0000000000000187): add tag: spam
[22:18:33.936603] [0x00007fe5e197f700] [debug] toggle_action: threadid:0000000000000186
[22:18:33.936686] [0x00007fe5e197f700] [debug] nm (0000000000000186): add tag: spam
terminate called after throwing an instance of 'Xapian::DatabaseCorruptError'
Abgebrochen (Speicherabzug geschrieben)
```

**Comment by moddx** [comment]
Although I waited several hours before I posted this potential bug, and  kept on validating that it does not work, now suddenly it does work just fine. Even for  50+ mails.

The only thing I changed in the meantime were the `save_sent_to`- and `save_drafts_to`-paths. Both contained a typo before.

I will check how this developes.

**Comment by gauteh** [comment]
Hi, any news on this? It does not immediately seem like an astroid bug. Do you mark the messages first? or just press toggle spam many times?

**Comment by moddx** [comment]
I marked a lot of messages with 't' and then flagged them as spam via ';' and 'S'.
You are probably right in that the bug is somewhere deeper down, e.g. in notmuch.

The bug did not occur again since then, although I had a few similar crashes when sending mail for example. But that is an entirely different topic.

**Comment by gauteh** [comment]
Ok, any crashes are pretty critical. I am going to close this one since we don't have a way to investigate or solve it at the moment. Please post logs and backtraces (gdb / core dumps) if possible.

**Comment by moddx** [comment]
Here are the core dumps, in case they contain anything useful.
The first core dump corresponds to the astroids log/console output from the first post.
The second core dump occured half a minute earlier, as I did the same thing (marking messages and then tagging them all as spam).

**Core Dump #1** (corresponds to the debug-log of the inital post)
```
Feb 08 22:18:34 archmail systemd-coredump[18107]: Process 18099 (astroid) of user 1000 dumped core.
                      
	  Stack trace of thread 18102:
	  #0  0x00007fe5f4ae704f raise (libc.so.6)
	  #1  0x00007fe5f4ae847a abort (libc.so.6)
	  #2  0x00007fe5f53174ed _ZN9__gnu_cxx27__verbose_terminate_handlerEv (libstdc++.so.6)
	  #3  0x00007fe5f53152a6 _ZN10__cxxabiv111__terminateEPFvvE (libstdc++.so.6)
	  #4  0x00007fe5f53152f1 _ZSt9terminatev (libstdc++.so.6)
	  #5  0x00007fe5f53415be execute_native_thread_routine (libstdc++.so.6)
	  #6  0x00007fe5f4e59454 start_thread (libpthread.so.0)
	  #7  0x00007fe5f4b9c7df __clone (libc.so.6)
	  
	  Stack trace of thread 18100:
	  #0  0x00007fe5f4b9348d poll (libc.so.6)
	  #1  0x00007fe5f61f4786 n/a (libglib-2.0.so.0)
	  #2  0x00007fe5f61f489c g_main_context_iteration (libglib-2.0.so.0)
	  #3  0x00007fe5f61f48e1 n/a (libglib-2.0.so.0)
	  #4  0x00007fe5f621c0d5 n/a (libglib-2.0.so.0)
	  #5  0x00007fe5f4e59454 start_thread (libpthread.so.0)
	  #6  0x00007fe5f4b9c7df __clone (libc.so.6)
	  
	  Stack trace of thread 18101:
	  #0  0x00007fe5f4b9348d poll (libc.so.6)
	  #1  0x00007fe5f61f4786 n/a (libglib-2.0.so.0)
	  #2  0x00007fe5f61f4b12 g_main_loop_run (libglib-2.0.so.0)
	  #3  0x00007fe5f67da316 n/a (libgio-2.0.so.0)
	  #4  0x00007fe5f621c0d5 n/a (libglib-2.0.so.0)
	  #5  0x00007fe5f4e59454 start_thread (libpthread.so.0)
	  #6  0x00007fe5f4b9c7df __clone (libc.so.6)
	  
	  Stack trace of thread 18104:
	  #0  0x00007fe5f4b9348d poll (libc.so.6)
	  #1  0x00007fe5f61f4786 n/a (libglib-2.0.so.0)
	  #2  0x00007fe5f61f489c g_main_context_iteration (libglib-2.0.so.0)
	  #3  0x00007fe5e00134bd n/a (libdconfsettings.so)
	  #4  0x00007fe5f621c0d5 n/a (libglib-2.0.so.0)
	  #5  0x00007fe5f4e59454 start_thread (libpthread.so.0)
	  #6  0x00007fe5f4b9c7df __clone (libc.so.6)
	  
	  Stack trace of thread 18099:
	  #0  0x00007fe5f4b9348d poll (libc.so.6)
	  #1  0x00007fe5f61f4786 n/a (libglib-2.0.so.0)
	  #2  0x00007fe5f61f489c g_main_context_iteration (libglib-2.0.so.0)
	  #3  0x00007fe5f67ae54d g_application_run (libgio-2.0.so.0)
	  #4  0x000000000047c742 _ZN7Astroid7Astroid4mainEiPPc (astroid)
	  #5  0x00007fe5f4ad4291 __libc_start_main (libc.so.6)
	  #6  0x000000000047e28a _start (astroid)
```

**Core dump #2** (same problem: mark -> tag all as spam -> crash)
```
Feb 08 22:17:54 archmail offlineimap[324]:  *** Finished account 'mail@host' in 0:01
Feb 08 22:17:54 archmail offlineimap[324]:  Next refresh in 1.0 minutes
Feb 08 22:18:01 archmail systemd-coredump[18083]: Process 16539 (astroid) of user 1000 dumped core.
                                                  
	  Stack trace of thread 16542:
	  #0  0x00007fdc9e3fd04f raise (libc.so.6)
	  #1  0x00007fdc9e3fe47a abort (libc.so.6)
	  #2  0x00007fdc9ec2d4ed _ZN9__gnu_cxx27__verbose_terminate_handlerEv (libstdc++.so.6)
	  #3  0x00007fdc9ec2b2a6 _ZN10__cxxabiv111__terminateEPFvvE (libstdc++.so.6)
	  #4  0x00007fdc9ec2b2f1 _ZSt9terminatev (libstdc++.so.6)
	  #5  0x00007fdc9ec575be execute_native_thread_routine (libstdc++.so.6)
	  #6  0x00007fdc9e76f454 start_thread (libpthread.so.0)
	  #7  0x00007fdc9e4b27df __clone (libc.so.6)
	  
	  Stack trace of thread 16739:
	  #0  0x00007fdc9e77510f pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0)
	  #1  0x00007fdc9ec514cc __gthread_cond_wait (libstdc++.so.6)
	  #2  0x00007fdc9be3723a _ZN3JSC14BlockAllocator22blockFreeingThreadMainEv (libjavascriptcoregtk-3.0.so.0)
	  #3  0x00007fdc9c12570a n/a (libjavascriptcoregtk-3.0.so.0)
	  #4  0x00007fdc9e76f454 start_thread (libpthread.so.0)
	  #5  0x00007fdc9e4b27df __clone (libc.so.6)
	  
	  Stack trace of thread 16740:
	  #0  0x00007fdc9e77510f pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0)
	  #1  0x00007fdc9ec514cc __gthread_cond_wait (libstdc++.so.6)
	  #2  0x00007fdc9be3866b _ZN3JSC8GCThread16waitForNextPhaseEv (libjavascriptcoregtk-3.0.so.0)
	  #3  0x00007fdc9be38758 _ZN3JSC8GCThread12gcThreadMainEv (libjavascriptcoregtk-3.0.so.0)
	  #4  0x00007fdc9c12570a n/a (libjavascriptcoregtk-3.0.so.0)
	  #5  0x00007fdc9e76f454 start_thread (libpthread.so.0)
	  #6  0x00007fdc9e4b27df __clone (libc.so.6)
	  
	  Stack trace of thread 16916:
	  #0  0x00007fdc9e4ab0b3 __select (libc.so.6)
	  #1  0x00007fdc9fb51e24 g_spawn_sync (libglib-2.0.so.0)
	  #2  0x00007fdca388cac8 _ZN4Glib10spawn_syncERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKNS_11ArrayHandleIS5_NS_17Container_Helpers10TypeTraitsIS5_EEEENS_10SpawnFlagsERKN4sigc4slotIvNSG_3nilESI_SI_SI_SI_SI_SI_EEPS5_SM_Pi (libglibmm-2.4.so.1)
	  #3  0x000000000059c5dc _ZN7Astroid10ThreadView12do_open_linkEN4Glib7ustringE (astroid)
	  #4  0x00000000005b1b70 _ZN4sigc8internal10slot_call0INS_12bind_functorILin1ENS_18bound_mem_functor1IvN7Astroid10ThreadViewEN4Glib7ustringEEES7_NS_3nilES9_S9_S9_S9_S9_EEvE7call_itEPNS0_8slot_repE (astroid)
	  #5  0x00007fdca388d52d n/a (libglibmm-2.4.so.1)
	  #6  0x00007fdc9fb320d5 n/a (libglib-2.0.so.0)
	  #7  0x00007fdc9e76f454 start_thread (libpthread.so.0)
	  #8  0x00007fdc9e4b27df __clone (libc.so.6)
	  
	  Stack trace of thread 16736:
	  #0  0x00007fdc9e77510f pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0)
	  #1  0x00007fdca1af1e4c n/a (libwebkitgtk-3.0.so.0)
	  #2  0x00007fdca1af34c5 n/a (libwebkitgtk-3.0.so.0)
	  #3  0x00007fdc9c12570a n/a (libjavascriptcoregtk-3.0.so.0)
	  #4  0x00007fdc9e76f454 start_thread (libpthread.so.0)
	  #5  0x00007fdc9e4b27df __clone (libc.so.6)
	  
	  Stack trace of thread 16735:
	  #0  0x00007fdc9e77510f pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0)
	  #1  0x00007fdc9c10cf26 _ZN3WTF17TCMalloc_PageHeap15scavengerThreadEv (libjavascriptcoregtk-3.0.so.0)
	  #2  0x00007fdc9c10cf49 _ZN3WTF17TCMalloc_PageHeap18runScavengerThreadEPv (libjavascriptcoregtk-3.0.so.0)
	  #3  0x00007fdc9e76f454 start_thread (libpthread.so.0)
	  #4  0x00007fdc9e4b27df __clone (libc.so.6)
	  
	  Stack trace of thread 16544:
	  #0  0x00007fdc9e4a948d poll (libc.so.6)
	  #1  0x00007fdc9fb0a786 n/a (libglib-2.0.so.0)
	  #2  0x00007fdc9fb0a89c g_main_context_iteration (libglib-2.0.so.0)
	  #3  0x00007fdc899294bd n/a (libdconfsettings.so)
	  #4  0x00007fdc9fb320d5 n/a (libglib-2.0.so.0)
	  #5  0x00007fdc9e76f454 start_thread (libpthread.so.0)
	  #6  0x00007fdc9e4b27df __clone (libc.so.6)
	  
	  Stack trace of thread 16541:
	  #0  0x00007fdc9e4a948d poll (libc.so.6)
	  #1  0x00007fdc9fb0a786 n/a (libglib-2.0.so.0)
	  #2  0x00007fdc9fb0ab12 g_main_loop_run (libglib-2.0.so.0)
	  #3  0x00007fdca00f0316 n/a (libgio-2.0.so.0)
	  #4  0x00007fdc9fb320d5 n/a (libglib-2.0.so.0)
	  #5  0x00007fdc9e76f454 start_thread (libpthread.so.0)
	  #6  0x00007fdc9e4b27df __clone (libc.so.6)
	  
	  Stack trace of thread 16540:
	  #0  0x00007fdc9e4a948d poll (libc.so.6)
	  #1  0x00007fdc9fb0a786 n/a (libglib-2.0.so.0)
	  #2  0x00007fdc9fb0a89c g_main_context_iteration (libglib-2.0.so.0)
	  #3  0x00007fdc9fb0a8e1 n/a (libglib-2.0.so.0)
	  #4  0x00007fdc9fb320d5 n/a (libglib-2.0.so.0)
	  #5  0x00007fdc9e76f454 start_thread (libpthread.so.0)
	  #6  0x00007fdc9e4b27df __clone (libc.so.6)
	  
	  Stack trace of thread 16539:
	  #0  0x00007fdc9e4a948d poll (libc.so.6)
	  #1  0x00007fdc9fb0a786 n/a (libglib-2.0.so.0)
	  #2  0x00007fdc9fb0a89c g_main_context_iteration (libglib-2.0.so.0)
	  #3  0x00007fdca00c454d g_application_run (libgio-2.0.so.0)
	  #4  0x000000000047c742 _ZN7Astroid7Astroid4mainEiPPc (astroid)
	  #5  0x00007fdc9e3ea291 __libc_start_main (libc.so.6)
	  #6  0x000000000047e28a _start (astroid)
```


As a bonus, here is another core dump from a crash that happend today, while opening an email (which also contained html and embedded images). Astroid crashed 1-2 seconds after rendering the mail.
As you can see OfflineImap finished it sync just seconds before the core dump, just like in core dump **#2**.  There was  however no activity from OfflineImap in the time of core dump **#1**, though that doesn't have to say anything, since  the middleman is notmuch anyway.


**Core dump #3** (while opening an email)
```
Feb 15 16:16:40 archmail offlineimap[343]:  *** Finished account 'mail@host' in 0:04
Feb 15 16:16:40 archmail offlineimap[343]:  Next refresh in 1.0 minutes
Feb 15 16:16:43 archmail systemd-coredump[783]: Process 417 (astroid) of user 1000 dumped core.
	
	Stack trace of thread 417:
	#0  0x00007f24547bb7a5 _ZN4Glib16DispatchNotifier17send_notificationEPNS_10DispatcherE (libglibmm-2.4.so.1)
	#1  0x000000000053a7f7 _ZN7Astroid4Mode9set_labelEN4Glib7ustringE (astroid)
	#2  0x0000000000566fb2 _ZN7Astroid11ThreadIndex14on_stats_readyEv (astroid)
	#3  0x00007f24547bbc02 _ZN4Glib16DispatchNotifier15pipe_io_handlerENS_11IOConditionE (libglibmm-2.4.so.1)
	#4  0x00007f24547be907 _ZN4Glib8IOSource8dispatchEPN4sigc9slot_baseE (libglibmm-2.4.so.1)
	#5  0x00007f24547be30f _ZN4Glib6Source14dispatch_vfuncEP8_GSourcePFiPvES3_ (libglibmm-2.4.so.1)
	#6  0x00007f2450a3243a g_main_context_dispatch (libglib-2.0.so.0)
	#7  0x00007f2450a327f0 n/a (libglib-2.0.so.0)
	#8  0x00007f2450a3289c g_main_context_iteration (libglib-2.0.so.0)
	#9  0x00007f2450fec54d g_application_run (libgio-2.0.so.0)
	#10 0x000000000047c742 _ZN7Astroid7Astroid4mainEiPPc (astroid)
	#11 0x00007f244f312291 __libc_start_main (libc.so.6)
	#12 0x000000000047e28a _start (astroid)
	
	Stack trace of thread 530:
	#0  0x00007f244f3d148d poll (libc.so.6)
	#1  0x00007f2450a32786 n/a (libglib-2.0.so.0)
	#2  0x00007f2450a32b12 g_main_loop_run (libglib-2.0.so.0)
	#3  0x00007f2451018316 n/a (libgio-2.0.so.0)
	#4  0x00007f2450a5a0d5 n/a (libglib-2.0.so.0)
	#5  0x00007f244f697454 start_thread (libpthread.so.0)
	#6  0x00007f244f3da7df __clone (libc.so.6)
	
	Stack trace of thread 781:
	#0  0x00007f244f69d10f pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0)
	#1  0x00007f244fb794cc __gthread_cond_wait (libstdc++.so.6)
	#2  0x00007f244cd6066b _ZN3JSC8GCThread16waitForNextPhaseEv (libjavascriptcoregtk-3.0.so.0)
	#3  0x00007f244cd60758 _ZN3JSC8GCThread12gcThreadMainEv (libjavascriptcoregtk-3.0.so.0)
	#4  0x00007f244d04d70a n/a (libjavascriptcoregtk-3.0.so.0)
	#5  0x00007f244f697454 start_thread (libpthread.so.0)
	#6  0x00007f244f3da7df __clone (libc.so.6)
	
	Stack trace of thread 529:
	#0  0x00007f244f3d148d poll (libc.so.6)
	#1  0x00007f2450a32786 n/a (libglib-2.0.so.0)
	#2  0x00007f2450a3289c g_main_context_iteration (libglib-2.0.so.0)
	#3  0x00007f2450a328e1 n/a (libglib-2.0.so.0)
	#4  0x00007f2450a5a0d5 n/a (libglib-2.0.so.0)
	#5  0x00007f244f697454 start_thread (libpthread.so.0)
	#6  0x00007f244f3da7df __clone (libc.so.6)
	
	Stack trace of thread 559:
	#0  0x00007f244f69d10f pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0)
	#1  0x00007f244fb794cc __gthread_cond_wait (libstdc++.so.6)
	#2  0x00000000005c11b6 _ZN7Astroid13ActionManager13action_workerEv (astroid)
	#3  0x00007f244fb7f58f execute_native_thread_routine (libstdc++.so.6)
	#4  0x00007f244f697454 start_thread (libpthread.so.0)
	#5  0x00007f244f3da7df __clone (libc.so.6)
	
	Stack trace of thread 562:
	#0  0x00007f244f3d148d poll (libc.so.6)
	#1  0x00007f2450a32786 n/a (libglib-2.0.so.0)
	#2  0x00007f2450a3289c g_main_context_iteration (libglib-2.0.so.0)
	#3  0x00007f24363144bd n/a (libdconfsettings.so)
	#4  0x00007f2450a5a0d5 n/a (libglib-2.0.so.0)
	#5  0x00007f244f697454 start_thread (libpthread.so.0)
	#6  0x00007f244f3da7df __clone (libc.so.6)
	
	Stack trace of thread 776:
	#0  0x00007f244f69d10f pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0)
	#1  0x00007f244d034f26 _ZN3WTF17TCMalloc_PageHeap15scavengerThreadEv (libjavascriptcoregtk-3.0.so.0)
	#2  0x00007f244d034f49 _ZN3WTF17TCMalloc_PageHeap18runScavengerThreadEPv (libjavascriptcoregtk-3.0.so.0)
	#3  0x00007f244f697454 start_thread (libpthread.so.0)
	#4  0x00007f244f3da7df __clone (libc.so.6)
	
	Stack trace of thread 777:
	#0  0x00007f244f69d10f pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0)
	#1  0x00007f2452a19e4c n/a (libwebkitgtk-3.0.so.0)
	#2  0x00007f2452a1b4c5 n/a (libwebkitgtk-3.0.so.0)
	#3  0x00007f244d04d70a n/a (libjavascriptcoregtk-3.0.so.0)
	#4  0x00007f244f697454 start_thread (libpthread.so.0)
	#5  0x00007f244f3da7df __clone (libc.so.6)
	
	Stack trace of thread 778:
	#0  0x00007f244f3d5f19 syscall (libc.so.6)
	#1  0x00007f2450a7803a g_cond_wait_until (libglib-2.0.so.0)
	#2  0x00007f2450a06e89 n/a (libglib-2.0.so.0)
	#3  0x00007f2450a5aaa6 n/a (libglib-2.0.so.0)
	#4  0x00007f2450a5a0d5 n/a (libglib-2.0.so.0)
	#5  0x00007f244f697454 start_thread (libpthread.so.0)
	#6  0x00007f244f3da7df __clone (libc.so.6)
	
	Stack trace of thread 779:
	#0  0x00007f244f3d5f19 syscall (libc.so.6)
	#1  0x00007f2450a7803a g_cond_wait_until (libglib-2.0.so.0)
	#2  0x00007f2450a06e89 n/a (libglib-2.0.so.0)
	#3  0x00007f2450a5aaa6 n/a (libglib-2.0.so.0)
	#4  0x00007f2450a5a0d5 n/a (libglib-2.0.so.0)
	#5  0x00007f244f697454 start_thread (libpthread.so.0)
	#6  0x00007f244f3da7df __clone (libc.so.6)
	
	Stack trace of thread 780:
	#0  0x00007f244f69d4b8 pthread_cond_timedwait@@GLIBC_2.3.2 (libpthread.so.0)
	#1  0x00007f244cd5f0a8 _ZN3JSC14BlockAllocator15waitForDurationENSt6chrono8durationIlSt5ratioILl1ELl1000EEEE (libjavascriptcoregtk-3.0.so.0)
	#2  0x00007f244cd5f17d _ZN3JSC14BlockAllocator22blockFreeingThreadMainEv (libjavascriptcoregtk-3.0.so.0)
	#3  0x00007f244d04d70a n/a (libjavascriptcoregtk-3.0.so.0)
	#4  0x00007f244f697454 start_thread (libpthread.so.0)
	#5  0x00007f244f3da7df __clone (libc.so.6)
```

**Comment by gauteh** [comment]
I am wondering if this is related to: https://github.com/astroidmail/astroid/issues/64, could this crash be happening when `notmuch new` is running somewhere in the background? Do you have a polling setup which makes notmuch new run very often?

**Comment by gauteh** [comment]
(By the way, for some reason I don't get e-mail notifications for this issue..)

**Comment by gauteh** [comment]
I think I have the error demonstrated in the test/test_notmuch_standalone script.

**Comment by moddx** [comment]
Yes I think I poll quite often.
OfflineImap should execute `notmuch new` every minute, because it does a quick sync every minute and a full sync every 10 minutes.

This is an excerpt from my .offlineimaprc

```
[Account someAccount]
..
autorefresh = 1.0
quick = 9
postsynchook = notmuch new
```

Now that I think about it, I may reduce the polling rate a notch.

**Comment by gauteh** [comment]
Note that astroid is designed to run offlineimap and notmuch new from within, since it blocks write access to the db internally for this. Have a look here: https://github.com/astroidmail/astroid/issues/276

**Comment by gauteh** [comment]
As mentioned in #276, we could facilitate some commands that would make external sync and notmuch new work better. Astroid needs to know that the database is being changed, and there is currently no notmuch way to do this. The commands I am thinking of are:

* `astroid --start-external-poll` - tells astroid that there is a poll going on (offlineimap + notmuch new), considers the db to be RW locked, saves the current db revision
* `astroid --stop-external-poll` - tells astroid the poll is done and the DB is free, reloads any messages since the stored db revision
* `astroid --refresh [revision or query]` - utility command that tells astroid to reload any messages since this revision or matching this query.

**Comment by moddx** [comment]
I don't necessarily need external polling. I only used it because the astroid docs were not very clear on its current disadvantages. The reason I used it is that OfflineImap, when used externaly, can keep an open connection to the mail server.

I just switched to astroid-initiated polling as you linked. It works quite well.

The commands you envision sound good  in general, but relying on the user to signal the stop of external polling, under all error conditions, seems like asking for trouble. Maybe astroid could indicate when a poll takes suspiciously long, via the ui. That way users, who use risky scripts, might get less irritated by mail not showing up in astroid, and have an indication on what might be wrong.
